### PR TITLE
Handle stripe api errors

### PR DESCRIPTION
**Pull request by SeriousMatt**

### CODE REVIEW

File: stripe-node/app.js
Comment (mcdurdin) [inline stripe-node/app.js]: Is `customers` an array? Why the plural?

File: stripe-node/app.js
Comment (mcdurdin) [inline stripe-node/app.js]: A card can be declined for many reasons. A message should be provided from the bank, it seems in the `message` property already?

Actually, I'm not sure why you are not just using the `message` property and instead are overwriting it in the error handler?

File: stripe-node/app.js
Comment (mcdurdin) [inline stripe-node/app.js]: Debug code?

File: stripe-node/app.js
Comment (mcdurdin) [inline stripe-node/app.js]: This needs to be inside the subsequent `if` statement

File: stripe-node/app.js
Comment (mcdurdin) [inline stripe-node/app.js]: This should probably be
```suggestion
      if (typeof err === 'object') {
```

File: stripe-node/app.js
Comment (mcdurdin) [inline stripe-node/app.js]: I would prefer to see error messages sent in some structured form (JSON?) from an API, rather than as plain text. Also, do you just send a HTTP/200 back to the website? Don't we need some sort of extra data, e.g. receipt #, etc?

**Review by mcdurdin — CHANGES_REQUESTED** [review]

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: good point. I'll leave it as customer.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: Yes. I'll take that out.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: This is what stripe provided for error handling. See https://stripe.com/docs/api/errors/handling I'll see about passing in the error message.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: The json response is definitely doable. I'll see about working within the solution provided by Stripe (https://stripe.com/docs/api/errors/handling) and sending back json. So as far as the receipt is concerned, that depends on how we are sending receipts. Stripe integrates and sends receipts on behalf of the data collected (which you should have a copy of that example receipt).  If we are sending the receipts manually, I agree this data would need to be sent either back to Angular for download.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: why is that?

File: stripe-node/app.js
Comment (mcdurdin) [reply stripe-node/app.js]: So according to the [Stripe docs](https://stripe.com/docs/api/errors#errors-message), `message` is passed back from Stripe and for card errors is appropriate to display to end users. Other messages are apparently not useful for end users.

File: stripe-node/app.js
Comment (mcdurdin) [reply stripe-node/app.js]: If `err` is not an `object`, you'll get an error: `TypeError: Cannot read property 'type' of ...`.

File: stripe-node/app.js
Comment (mcdurdin) [reply stripe-node/app.js]: Something needs to be sent back to the user. If they mistype their email address, etc, then some reference number is important.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: That makes sense. So the user/application won't know if any of the errors listed unless a response is sent back and any error info is helpful. It's better than having no error information at all.

Comment (SeriousMatt) [reply stripe-node/app.js]: I'll go ahead set it back to err.message. Think it would be good to set that for all of them.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: This should be resolved since that console.log was removed.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: Your proposed solution still results in an error when there is no error occurring. I've left it for now since it is working.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: Can you clarify on what you mean by "if they mistype their email address, etc"

**Comment by SeriousMatt** [comment]
So just a note on how I am working on this project so far (mainly for a point of clarity). I am opening up a pull request initially so we can discuss changes and see the work progress as it comes along and then when I am satisfied with it, I am requesting review from you (perhaps that was done prematurely on my part this time?). At that point, feel free to take a look then because I will be waiting for you to approve it before it gets merged. Of course, you are free to look at the code at any point and make suggestions (i.e. hold me accountable...) love the feedback 🥇. But fyi, some of what you addressed was not actually part of the final pull request. Just want to make sure that your time is well-spent and you are not necessarily seeing each commit as pull-request ready for a production merge @mcdurdin. Thanks!

File: stripe-node/app.js
Comment (mcdurdin) [reply stripe-node/app.js]: Okay. Imagine the user goes through the process of making a donation. But instead of typing `marc@example.com`, they type `mar@example.com`. This would mean they would never receive the emailed receipt. What would they see on the screen in that instance with your current design? Would they see any receipt number, or other reference number that they would be able to contact us with -- seeing as we would not be able to contact them?

**Comment by mcdurdin** [comment]
For a pull-request that isn't ready for review, in our workflow we label it as 'work-in-progress' and add a '[WIP]' prefix to the title. I'll do that to this PR.

**Comment by mcdurdin** [comment]
Also, I imagined once you asked me to review that it was ready for review?

-- Sorry -- just saw the order of request-for-review and realised that your request to review this came in after your changes. So I won't add the [WIP] title in etc at this point.

File: stripe-node/api/charge.js
Comment (mcdurdin) [inline stripe-node/api/charge.js:18]: Could you format this over multiple lines? Makes readability better in code reviews etc.

**Review by mcdurdin — APPROVED** [review]

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: Makes sense. I will work on getting this added (sending back receipt number).

File: stripe-node/api/charge.js
Comment (SeriousMatt) [reply stripe-node/api/charge.js:18]: I thought the same thing, spreading it proved problematic. Will try a few different options.

File: stripe-node/app.js
Comment (SeriousMatt) [reply stripe-node/app.js]: ^^ *in-the-next-pr*

**Comment by mcdurdin** [comment]
How about:

const description = 'Dear ' + req.body.data.billing_name +
', Thank you for your generous contribution of '+ req.body.currency + ' ' +
amount.toFixed(2) + ' to SIL International, with preference for ' +
req.body.data.billing_name  +'. This letter serves as a receipt for your ' +
'contribution. Founded over 80 years ago, SIL International is committed ' +
'to serving language communities worldwide as they build capacity for ' +
'sustainable language development. SIL does this primarily through ' +
'research, translation, training and materials development. SIL works ' +
'alongside ethnolinguistic communities and their partners as they discover '+
how language development addresses the challenging areas of their daily ' +
lives—social, cultural, political, economic and spiritual. Currently, SIL ' +
'works alongside speakers of more than 1,700 languages in over 100 '+
countries. SIL International is a nonprofit organization under section '+
501(c)(3) of the IRS code. No goods or services were provided in '+
consideration of this gift.  Your donation is tax deductible. Thank you '+
for partnering with us in making a difference in the lives of people all '+
'over the world!';

**Pull request merged**
